wayland-client/src/async_event_queue.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    io,
+    num::NonZeroUsize,
+    sync::{
+        Arc,
+        atomic::Ordering,
+    },
+    os::fd::{
+        AsFd,
+        BorrowedFd,
+    },
+};
+use crate::{
+    conn::SyncData,
+    Connection,
+    EventQueue,
+    DispatchError,
+    WaylandError,
+    protocol::{
+        wl_display,
+    },
+};
+use tokio::io::{
+    unix::AsyncFd,
+    Ready,
+};
+
+/// Asynchronous version of [`EventQueue`], taking borrowed references to an [`EventQueue`],
+/// allowing for non-blocking polled asynchronous operations against the Queue using
+/// [`AsyncFd`] from the tokio runtime to poll against the wayland socket
+#[derive(Debug)]
+pub struct AsyncEventQueue<'a, State> {
+    /// Inner reference to synchronous version of this event queue
+    pub queue: EventQueue<State>,
+    conn: &'a Connection,
+    afd: AsyncFd<BorrowedFd<'a>>,
+}
+
+impl<'a, State> AsyncEventQueue<'a, State> {
+    /// Creates a new [`AsyncEventQueue`] from a reference to a given [`Connection`] and
+    /// *synchronous* [`EventQueue`]
+    #[inline]
+    pub fn new(
+        conn: &'a Connection,
+        event_queue: EventQueue<State>,
+    ) -> io::Result<Self> {
+        Ok(AsyncEventQueue {
+            queue: event_queue,
+            conn: conn,
+            afd: AsyncFd::new(conn.as_fd())?,
+        })
+    }
+
+    #[inline(always)]
+    fn dispatch_pending(&mut self, data: &mut State) -> Result<Option<NonZeroUsize>, DispatchError> {
+        self.queue.dispatch_pending(data)
+            .map(NonZeroUsize::new)
+    }
+
+    /// Asynchronous version of [`EventQueue::blocking_dispatch`] that uses tokio's fd-based
```

Comment:
Doc comment needs adjusting due to API changes from reference -> owned version of `EventQueue` member.